File: backend_java_spring_boot/myPizza/src/main/java/backendapp/myPizza/services/OrderService.java
package backendapp.myPizza.services;

import backendapp.myPizza.Models.entities.*;
import backendapp.myPizza.Models.enums.OrderStatus;
import backendapp.myPizza.Models.reqDTO.OrderInitDTO;
import backendapp.myPizza.Models.reqDTO.OrderSetDTO;
import backendapp.myPizza.Models.reqDTO.SendOrderDTO;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.Models.resDTO.OrderCheckoutInfo;
import backendapp.myPizza.Models.resDTO.OrderInitRes;
import backendapp.myPizza.SocketIO.services.MessageService;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.exceptions.NotFoundException;
import backendapp.myPizza.exceptions.UnauthorizedException;
import backendapp.myPizza.repositories.*;
import backendapp.myPizza.security.JwtUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class OrderService {

    @Autowired
    private OrderRepository orderRp;

    @Autowired
    private OrderSetRepository orderSetRp;

    @Autowired
    private ProductRepository productRp;

    @Autowired
    private JwtUtils jwtUtils;

    @Autowired
    private UserRepository userRp;

    @Autowired
    private ToppingRefRepository toppingRefRp;

    @Autowired
    private ProductRefRepository productRefRp;

    @Autowired
    private _SessionService _session;

    @Autowired
    private TimeIntervalRepository timeIntervalRp;

    @Autowired
    private MessageService messageSvc;

    public OrderInitRes orderInit(OrderInitDTO orderInitDTO) throws BadRequestException {
        Order order = new Order();
        order.setStatus(OrderStatus.INIT);
        orderRp.save(order);
        if (orderInitDTO.orderSetsDTO().isEmpty()) throw new BadRequestException("Order sent is empty");
        List<OrderSet> orderSets = new ArrayList<>();
        for (OrderSetDTO o : orderInitDTO.orderSetsDTO()) {
            Product p = productRp.findById(o.productId()).orElseThrow(
                    () -> new BadRequestException("Product with id='" + o.productId() + "' doesn't exist")
            );
            p.setProductTotalAmount();
            if (o.quantity() <= 0)
                throw new BadRequestException("Product with id='" + o.productId() + "': quantity must be an integer number major than 0");
            ProductRef pr = new ProductRef(p.getName(), p.getPrice());
            for (Topping t : p.getToppings()) {
                ToppingRef tr = new ToppingRef(t.getName(), t.getPrice());
                toppingRefRp.save(tr);
                pr.getToppingsRef().add(tr);
            }
            productRefRp.save(pr);
            OrderSet orderSet = new OrderSet(pr, o.quantity());
            orderSet.setOrder(order);
            orderSets.add(orderSet);
        }
        orderSetRp.saveAll(orderSets);
        orderRp.save(order);
        return new OrderInitRes(order.getId(), order.getStatus());
    }

    public OrderCheckoutInfo getClientOrderInit(Boolean guest, Order order) throws UnauthorizedException, BadRequestException {
        if (guest) {
            // da implementare
            return null;
        } else {
            UUID userId = jwtUtils.extractUserIdFromReq();
            assert userRp.findById(userId).isPresent();
            User user = userRp.findById(userId).get();
            Address address = user.getAddresses().stream().filter(Address::is_default).findFirst()
                    .orElseThrow(
                            () -> new BadRequestException("Cannot find default address, please set a default address before retrying")
                    );
            OrderStatus status = order.getStatus();
            List<OrderSet> orderSets = order.getOrderSets();
            double deliveryCost = order.getDeliveryCost();
            double totalAmount = 0;
            for (OrderSet os : orderSets) {
                totalAmount += os.getQuantity() * os.getProductRef().getPrice();
            }
            totalAmount += deliveryCost;

            return new OrderCheckoutInfo(order.getId(), address, orderSets, status, deliveryCost, totalAmount);
        }

    }

    public ConfirmRes sendOrder(SendOrderDTO sendOrderDTO, boolean guest) throws BadRequestException, UnauthorizedException {
        Order order = orderRp.findById(sendOrderDTO.orderId()).orElseThrow(
                () -> new BadRequestException("Order you're trying to send doesn't exist")
        );
        if (!order.getStatus().equals(OrderStatus.INIT))
            throw new BadRequestException("An order must have INIT status to be sent");
        order.setStatus(OrderStatus.PENDING);
        UUID userId = guest ? jwtUtils.extractGuestUserIdFromReq() : jwtUtils.extractUserIdFromReq();
        User user = userRp.findById(userId).orElseThrow(
                () -> new BadRequestException("User doesn't exist")
        );
        order.setUser(user);
        order.setAddress(sendOrderDTO.address());
        order.setAsap(sendOrderDTO.asap());
        order.setExpectedDeliveryTime(sendOrderDTO.expectedDeliveryTime());
        order.setOrderTime(System.currentTimeMillis());
        orderRp.save(order);
        return new ConfirmRes("Order with id='" + sendOrderDTO.orderId() + " confirmed successfully", HttpStatus.OK);
    }

    public Order getOrderById(UUID orderId) throws BadRequestException {
        return orderRp.findById(orderId).orElseThrow(
                () -> new BadRequestException("order with id='" + orderId + "' doesn't exist")
        );
    }

    public ConfirmRes confirmOrder(UUID orderId, UUID timeIntervalId) throws BadRequestException {
        TimeInterval timeInterval = timeIntervalRp.findById(timeIntervalId).orElseThrow(
                () -> new BadRequestException("TimeInterval with id='" + timeIntervalId + "' doesn't exist")
        );
        Order order = orderRp.findById(orderId).orElseThrow(
                () -> new BadRequestException("Order with id='" + orderId + "' doesn't exist")
        );
        order.setTimeInterval(timeInterval);
        order.setStatus(OrderStatus.ACCEPTED);
        order.setDeliveryTime(timeInterval.getEndsAt());
        orderRp.save(order);

        return new ConfirmRes("order with id ='" + orderId + "' confirmed successfully", HttpStatus.OK);
    }



    public ConfirmRes rejectOrder(UUID orderId) throws BadRequestException {
        Order order = orderRp.findById(orderId).orElseThrow(
                () -> new BadRequestException("Order you're trying to reject doesn't exist")
        );
        if (!order.getStatus().equals(OrderStatus.PENDING))
            throw new BadRequestException("An order must have PENDING status to be confirmed or rejected");
        order.setStatus(OrderStatus.REJECTED);
        orderRp.save(order);
        return new ConfirmRes("Order with id='" + orderId + " rejected", HttpStatus.OK);
    }

    public List<TimeInterval> completeOrder(UUID orderId) throws BadRequestException, NotFoundException {
        Order order = orderRp.findById(orderId).orElseThrow(
                () -> new BadRequestException("Order you're trying to confirm doesn't exist")
        );
        if (!order.getStatus().equals(OrderStatus.ACCEPTED))
            throw new BadRequestException("An order must have ACCEPTED status to be confirmed or rejected");
        order.setStatus(OrderStatus.COMPLETED);
        order.setCompletedAt(System.currentTimeMillis());
        orderRp.save(order);
        return _session.getActiveSessionTimeIntervals();

    }


}

File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/order/services/order.service.ts
import { BadRequestException, HttpStatus, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Order } from '../entities/order.entity';
import { Repository } from 'typeorm';
import { JwtUtilsService } from 'src/nest_modules/auth-user/services/jwt-utils.service';
import { OrderSet } from '../entities/order-set.entity';
import { Product } from 'src/nest_modules/product/entities/product.entity';
import { ToppingRef } from '../entities/topping-ref.entity';
import { ProductRef } from '../entities/product-ref.entity';
import { TimeInterval } from '../entities/time-interval.entity';
import { OrderInitDTO } from '../interfaces/order-init-dto.interface';
import { OrderInitRes } from '../interfaces/order-init-res.interface';
import { OrderStatus } from '../enums/order-status.enum';
import { OrderCheckoutInfo } from '../interfaces/order-checkout-info.interface';
import { FastifyRequest } from 'fastify';
import { UUID } from 'crypto';
import { SendOrderDTO } from '../interfaces/send-order-dto.interface';
import { ConfirmRes } from 'src/nest_modules/auth-user/interfaces/confirm-res.interface';
import { OrderRes } from '../interfaces/order-res-interface';
import { SessionService } from './session.service';

@Injectable()
export class OrderService {

    constructor(
        private jwtUtils: JwtUtilsService,
        @InjectRepository(Order) private orderRepository: Repository<Order>,
        @InjectRepository(OrderSet) private orderSetRepository: Repository<OrderSet>,
        @InjectRepository(Product) private productRepository: Repository<Product>,
        @InjectRepository(ToppingRef) private toppingRefRepository: Repository<ToppingRef>,
        @InjectRepository(ProductRef) private productRefRepository: Repository<ProductRef>,
        @InjectRepository(TimeInterval) private timeIntervalrepository: Repository<TimeInterval>,
        private _session: SessionService
    ) { }

    public async orderInit(orderInitDTO: OrderInitDTO): Promise<OrderInitRes> {
        const order = new Order()
        order.status = OrderStatus.INIT
        await this.orderRepository.save(order)
        if (orderInitDTO.orderSetsDTO.length === 0)
            throw new BadRequestException('Order sent is empty')
        const ordersets: OrderSet[] = []
        for await (const orderSetDTO of orderInitDTO.orderSetsDTO) {
            const product = await this.productRepository.findOneBy({ id: orderSetDTO.productId })
            if (!product) {
                this.orderRepository.delete(order.id)
                throw new BadRequestException(`Product with id='${orderSetDTO.productId}' doesn't exist`)
            }
            if (orderSetDTO.quantity <= 0) {
                this.orderRepository.delete(order.id)
                throw new BadRequestException(`Product with id='${product.id}': quantity must be an integer number major than 0`)
            }
            product.setProductTotalAmount()
            const productRef = new ProductRef(product.name, product.price)
            productRef.toppingsRef = []
            if (product.toppings) {
                for await (const topping of product.toppings) {
                    const toppingRef = new ToppingRef(topping.name, topping.price)
                    await this.toppingRefRepository.save(toppingRef)
                    productRef.toppingsRef.push(toppingRef)
                }
            }
            await this.productRefRepository.save(productRef)
            const orderSet = new OrderSet(productRef, orderSetDTO.quantity)
            orderSet.order = order
            ordersets.push(orderSet)
        }
        await this.orderSetRepository.save(ordersets)
        await this.orderRepository.save(order)

        return {
            orderId: order.id,
            status: order.status
        }
    }

    public async getClientOrderInit(order: Order, req: FastifyRequest, guest?: boolean): Promise<OrderCheckoutInfo> {
        if (guest) {
            // da implementare
            return null
        } else {
            if (order.status !== OrderStatus.INIT) throw new BadRequestException('Order status must be INIT')
            const user = await this.jwtUtils.getUserFromReq(req)
            const address = user.addresses.find(address => address._default)
            if (!address) throw new BadRequestException('Cannot find default address, please set a default address before retrying')
            const totalAmount = order.orderSets.reduce((c, p) => c + p.productRef.price, 0) + order.deliveryCost
            return {
                orderId: order.id,
                address,
                orderSets: order.orderSets,
                status: order.status,
                deliveryCost: order.deliveryCost,
                totalAmount
            }
        }
    }

    private generateOrderResModel(order: Order): OrderRes {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { id, timeInterval, ...otherProps } = order
        return {
            id,
            ...otherProps
        }
    }

    public async getOrderById(id: UUID): Promise<OrderRes> {
        const order = await this.orderRepository.findOneBy({ id })
        if (!order) throw new NotFoundException(`Order with id='${id}' not found`)
        return this.generateOrderResModel(order)
    }

    public async sendOrder(sendOrderDTO: SendOrderDTO, req: FastifyRequest, /* guest?: boolean */): Promise<ConfirmRes> {
        const order = await this.orderRepository.findOneBy({ id: sendOrderDTO.orderId })
        if (!order) throw new BadRequestException("Order you're trying to send doesn't exist")
        if (order.status !== OrderStatus.INIT) // controllo da implementare
            throw new BadRequestException("An order must have INIT status to be sent")
        order.status = OrderStatus.PENDING
        const user = await this.jwtUtils.getUserFromReq(req)
        order.user = user
        order.address = sendOrderDTO.address
        order.asap = sendOrderDTO.asap
        order.expectedDeliveryTime = sendOrderDTO.expectedDeliveryTime
        await this.orderRepository.save(order)
        return {
            message: "Order with id='" + sendOrderDTO.orderId + " confirmed successfully",
            statusCode: HttpStatus.OK,
            timestamp: new Date().getTime()
        }

    }

    public async confirmOrder(orderId: UUID, timeIntervalId: UUID): Promise<ConfirmRes> {
        const timeInterval = await this.timeIntervalrepository.findOneBy({ id: timeIntervalId })
        if (!timeInterval) throw new BadRequestException("TimeInterval with id='" + timeIntervalId + "' doesn't exist")
        const order = await this.orderRepository.findOneBy({ id: orderId })
        if (!order) throw new BadRequestException("Order with id='" + orderId + "' doesn't exist")
        order.timeInterval = timeInterval
        order.status = OrderStatus.ACCEPTED
        order.deliveryTime = timeInterval.endsAt
        await this.orderRepository.save(order)
        return {
            message: "order with id ='" + orderId + "' confirmed successfully",
            timestamp: new Date().getTime(),
            statusCode: HttpStatus.OK
        }
    }

    public async rejectOrder(orderId: UUID): Promise<ConfirmRes> {
        const order = await this.orderRepository.findOneBy({ id: orderId })
        if (!order) throw new BadRequestException("Order with id='" + orderId + "' doesn't exist")
        if (order.status !== OrderStatus.PENDING)
            throw new BadRequestException('An order must have PENDING status to be confirmed or rejected')
        order.status = OrderStatus.REJECTED
        await this.orderRepository.save(order)
        return {
            message: "Order with id='" + orderId + " rejected",
            timestamp: new Date().getTime(),
            statusCode: HttpStatus.OK
        }
    }

    public async completeOrder(orderId: UUID): Promise<TimeInterval[]> {
        const order = await this.orderRepository.findOneBy({ id: orderId })
        if (!order) throw new BadRequestException("Order with id='" + orderId + "' doesn't exist")
        if (order.status !== OrderStatus.ACCEPTED)
            throw new BadRequestException('An order must have ACCEPTED status to be set as completed')
        order.status = OrderStatus.COMPLETED
        order.completedAt = new Date().getTime()
        await this.orderRepository.save(order)
        return await this._session.getActiveSessionTimeIntervals()
    }


}
